### PR TITLE
!!! TASK: Rename `_hiddenInIndex` to `hiddenInMenu`

### PR DESCRIPTION
**Pull request by kitsunet**
Renames `_hiddenInIndex` to `hiddenInMenu`.

This was done and decided in order to minimise confusion regarding magic `_*` underscore properties.
The property `_hiddenInIndex` _was_ such underscore property in 8.3 but as with Neos 9 `Node::isHiddenInIndex` has been removed. As non breaking temporary replacement it was reimplemented as _actual_ node property except with an underscore prefix: `_hiddenInIndex`. This is possible as the core doesn't work with these underscore properties anymore.
In Neos 9 there will simply be no magic meaning for underscore properties anymore, any property will now be allowed to start with an underscore. To better communicate this - but not enforce the use of such - Neos.Neos will refrain from declaring Node properties with underscores as this is an overloaded feature of the past and can easily cause confusion.

Fixes partially: #4208

**Upgrade instructions**

Following Fusion prototypes no longer accept an `renderHiddenInIndex` option but only consider `renderHiddenInMenu`. In the case you overwrote the default behaviour for any of these prototypes you should rename the option to `renderHiddenInMenu`.

- Neos.Neos:BreadcrumbMenu
- Neos.Neos:BreadcrumbMenuItems
- Neos.Neos:DimensionsMenu
- Neos.Neos:DimensionsMenuItems
- Neos.Neos:Menu
- Neos.Neos:MenuItems

Accessing the property has to be adjusted as well.

Upgrade from 8.3

```diff
- $node->isHiddenInIndex();
+ $node->getProperty('hiddenInMenu');
```

upgrade from 9.x-dev

```diff
- $node->getProperty('_hiddenInIndex');
+ $node->getProperty('hiddenInMenu');
```

**Review instructions**

### CODE REVIEW

**Comment by kitsunet** [comment]
unrelated errors 🤷

**Comment by bwaidelich** [comment]
> unrelated errors 🤷

@kitsunet Fixed with #4924 and rebased

File: Neos.Neos/Classes/Fusion/Helper/NodeHelper.php [hunk 1]
```diff
@@ -156,6 +157,17 @@ public function subgraphForNode(Node $node): ContentSubgraphInterface
         return $this->contentRepositoryRegistry->subgraphForNode($node);
     }
 
+    public function isNodeStateHidden(Node $node): bool
```
Comment (bwaidelich) [inline Neos.Neos/Classes/Fusion/Helper/NodeHelper.php]: I just stumbled upon the name of this method. Is the _state_ of the node hidden? And how is this related to the change?

**Review by bwaidelich — COMMENTED** [review]
Just one comment for now. Apart from that, the PR looks great!
You mentioned that there are follow-up todos for Neos UI?

**Comment by ahaeslich** [comment]
> You mentioned that there are follow-up todos for Neos UI?

Looking at this PR I noticed `Neos.Seo` may also need adjustments. So checking our codebase I see that:

- the legacy migration should rewrite `hiddenInIndex` to `hiddenInMenu`
- translations must use the new name
- the UI uses `_hiddenInIndex`
- there are comments referencing the prop
- our rector package needs adjustments
- Neos.Demo needs adjustments in fusion and with the exported content

File: Neos.Neos/Classes/Fusion/Helper/NodeHelper.php (hunk 1 above)
Comment (kitsunet) [reply Neos.Neos/Classes/Fusion/Helper/NodeHelper.php]: I left it close to what it is called internally but AFAIK when we don't support property('_hidden') there would be no way to figure out if a node is hidden, so I added this helper :)

Comment (bwaidelich) [reply Neos.Neos/Classes/Fusion/Helper/NodeHelper.php]: But wouldn't `isNodeHidden()` be better since the _node_ is hidden not its state.
But more confusing to me is the fact that we decided to call it `hiddenInMenu` and `disabled` IIRC.
I also wonder whether we should introduce unused code. And with SubtreeTags (#4659) we can probably just use `node.tags.contain('disabled')`, right?

**Comment by kitsunet** [comment]
Rector btw. is here https://github.com/neos/rector/pull/45

File: Neos.Neos/Classes/Fusion/Helper/NodeHelper.php (hunk 1 above)
Comment (kitsunet) [reply Neos.Neos/Classes/Fusion/Helper/NodeHelper.php]: If it works that way with the tags would be neat!

I am totally happy with removing this or changing the naming, shouldn't we then internally also use disabled everywhere? I think the point with the hidden state (finder) was originally to make it easy to just know if this node should be treated as disabled/hidden regardless if it's self hidden or just by inheritance. That's what this was meant for.

**Comment by kitsunet** [comment]
- [x] the legacy migration should rewrite hiddenInIndex to hiddenInMenu
- [x] translations must use the new name

**Comment by kitsunet** [comment]
These testfailures seem again unreltaed?

**Comment by kitsunet** [comment]
ok interesting, those tests are from a recently merged PR in 9.0, but why do they show up here given that they are not in my branch?

Note this is because we get the "merge" PR branch in tests that is actually a representation of 9.0 HEAD + this PR branch and not just plainly this PR branch.

**Comment by kitsunet** [comment]
FYI: Karsten said just renaming the labels should be fine.

**Comment by kitsunet** [comment]
Neos.Demo: https://github.com/neos/Neos.Demo/pull/192

**Comment by nezaniel** [comment]
I'd like to help here but can't get PHPStorm to detect the merge conflicts...

**Comment by kitsunet** [comment]
I can try to get it back on track, somehow I thought iut was long merged 🙈

**Comment by kitsunet** [comment]
neos-ui change: https://github.com/neos/neos-ui/pull/3735 
this fixes disabled tag export as well

**Comment by nezaniel** [comment]
what parts are still missing here? All that I can see looks very nice already

**Comment by kitsunet** [comment]
> what parts are still missing here? All that I can see looks very nice already

Nothing IMHO, it should all be merged please 😆

File: Neos.ContentRepository.NodeAccess/Classes/FlowQueryOperations/FilterOperation.php [hunk 2]
```diff
@@ -122,11 +122,9 @@ protected function getPropertyPath($element, $propertyPath)
         if ($propertyPath === '_identifier') {
             // TODO: deprecated (Neos <9 case)
             return $element->nodeAggregateId->value;
-        } elseif ($propertyPath[0] === '_' && $propertyPath !== '_hiddenInIndex') {
-            return ObjectAccess::getPropertyPath($element, substr($propertyPath, 1));
```
Comment (mhsdesign) [inline Neos.ContentRepository.NodeAccess/Classes/FlowQueryOperations/FilterOperation.php:126]: move the removal of `ObjectAccess` in the filter operation into a new pr

File: Neos.Neos/Classes/Fusion/Helper/NodeHelper.php [hunk 3]
```diff
@@ -156,6 +157,11 @@ public function subgraphForNode(Node $node): ContentSubgraphInterface
         return $this->contentRepositoryRegistry->subgraphForNode($node);
     }
 
+    public function isDisabled(Node $node): bool
+    {
+        return $node->tags->contain(SubtreeTag::disabled());
+    }
```
Comment (mhsdesign) [inline Neos.Neos/Classes/Fusion/Helper/NodeHelper.php]: the addition of this utility might also be better of in a separate pr where we can explain its usage and think of migrations

**Review by mhsdesign — COMMENTED** [review]
Thanks for taking care.
To solve the issue https://github.com/neos/neos-development-collection/issues/4208 fully we have to get rid of `ObjectAccess` in the `property` operation as well and possibly also in other places.

But instead of adding this into this pr and bloating it up i would like to suggest to keep this pr super simple and only related to `_hiddenInIndex` -> `hiddenInMenu` that way its much more structured.

All the general underscore `_` logic could then be removed in one batch with another pr. What do you say? I might also be able to help with that.

**Review by nezaniel — APPROVED** [review]
LGTM, even with the unrelated changes

**Comment by mhsdesign** [comment]
Extracted everything i though this pr was originally meant to be into https://github.com/neos/neos-development-collection/pull/5015 and added the things i found missing.

This pr now is super slim and only addresses the `hiddenInMenu` part.
@kitsunet i think we might need a better description yet (also maybe for the WHY we did this).
Rector seems already to be adjusted: https://github.com/neos/rector/pull/45

Followups:
- [x] https://github.com/neos/Neos.Demo/pull/192
- [x] https://github.com/neos/neos-seo/pull/183
- [x] https://github.com/neos/neos-ui/pull/3735

**Review by mhsdesign — APPROVED** [review]
Can be merged after a nice description has been added ;)

File: Neos.Neos/Classes/Fusion/AbstractMenuItemsImplementation.php [hunk 4]
```diff
@@ -76,17 +76,17 @@ public function isCalculateItemStatesEnabled(): bool
...
      *
      * @return boolean
      */
-    public function getRenderHiddenInIndex()
+    public function getRenderHiddenInMenu()
     {
-        if ($this->renderHiddenInIndex === null) {
-            $this->renderHiddenInIndex = (bool)$this->fusionValue('renderHiddenInIndex');
+        if ($this->renderHiddenInMenu === null) {
+            $this->renderHiddenInMenu = (bool)($this->fusionValue('renderHiddenInMenu') ?? $this->fusionValue('renderHiddenInIndex'));
```
Comment (Benjamin-K) [inline Neos.Neos/Classes/Fusion/AbstractMenuItemsImplementation.php]: Do we really want to support the old name, too? Would be cleaner to have a code migration for that as otherwise people will still keep on using the old name and wonder, why there is no corresponding yaml definition.

Comment (mhsdesign) [reply Neos.Neos/Classes/Fusion/AbstractMenuItemsImplementation.php]: Hi ;) yes good catch ... we do have working migrations regarding eel but fusion property names cannot be that easily migrated ... thus a super slim b/c layer like this might not hurt?

But we should adjust the docs to reference this ;)

**Review by mhsdesign — CHANGES_REQUESTED** [review]
Oh i just found out that the docs must be adjusted as well:

https://github.com/neos/neos-development-collection/blob/7faa3c8d1a6b8fd73623212380faa9979470e4f8/Neos.Neos/Documentation/References/NeosFusionReference.rst?plain=1#L1110

And there are still a lot of references to `hiddenInIndex` (especially in fusion code, docs and old test code)

The test old legacy test code referencing this name will be removed via my pr here: https://github.com/neos/neos-development-collection/pull/5036

File: Neos.Neos/Classes/Fusion/AbstractMenuItemsImplementation.php (hunk 4 above)
Comment (Benjamin-K) [reply Neos.Neos/Classes/Fusion/AbstractMenuItemsImplementation.php]: Is there a way to show an error to developers for this? Sth like:
```
if ($this->fusionValue('renderHiddenInIndex') !== null) {
    throw new Exception(...);
}
```

I know, this has small performance drawbacks, but I would prefer a clean migration personally, as this could lead to confusion in the long run. And the warning can be removed in a later version.

Comment (mhsdesign) [reply Neos.Neos/Classes/Fusion/AbstractMenuItemsImplementation.php]: Okay yes we could go down this road and additionally consider doing a string replace in all fusion files ;)
Thats a good idea i agree. Wdyt @kitsunet?

Comment (kitsunet) [reply Neos.Neos/Classes/Fusion/AbstractMenuItemsImplementation.php]: I wouldn't mind that, I think here was some opposition against throwing errors in fusion implementations as usage can be pretty conditional and you might not notice it before bringing in live.

Comment (mhsdesign) [reply Neos.Neos/Classes/Fusion/AbstractMenuItemsImplementation.php]: Okay i just removed it without exceptions thrown ^^ lets see how we can migrate this 🙈 ?

**Comment by mhsdesign** [comment]
Okay i pushed a few commits adjusting fusion and documentation further. I think there are now no references left.
This is now even more breaking as there is no fallback layer in Fusion thus we need a good upgrade instruction in the pr and maybe rector?

**Review by mhsdesign — APPROVED** [review]

**Comment by kdambekalns** [comment]
This broke the XLIFF files, Weblate says `Opening and ending tag mismatch: body line 4 and trans-unit, line 20, column 20 (<string>, line 20)` for `Neos.Neos/Resources/Private/Translations/*/NodeTypes/Document.xlf`

**Comment by dlubitz** [comment]
Looks like to many `</trans-unit>` tags.

**Comment by kdambekalns** [comment]
Yep, exactly:
image

**Comment by mhsdesign** [comment]
Fyi we missed to include some followups in the beta10 and might also have to consider providing a core node migration or a snipped to be copied (especially looking at @ahaeslich and her project as well as the docs and @nezaniel starship?)

**Comment by ahaeslich** [comment]
> Fyi we missed to include some followups in the beta10 and might also have to consider providing a core node migration or a snipped to be copied (especially looking at @ahaeslich and her project as well as the docs and @nezaniel starship?)

🤔 Did you discuss a core migration for this in a recent weekly? I'm undecided here ... How about all those users testing the beta releases?
